File: Data Structures & Algorithms/trapping-rain-water/submission-1.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;

        int[] maxLeft = new int[n];
        int[] maxRight = new int[n];

        for(int i = 1; i < n; i++)
            maxLeft[i] = Math.max(height[i-1], maxLeft[i - 1]);

        for(int i = n - 2; i >= 0; i--)
            maxRight[i] = Math.max(height[i + 1], maxRight[i + 1]);

        int maxArea = 0;
        for(int i = 0; i < n; i++) {
            int currArea = Math.min(maxLeft[i], maxRight[i]) - height[i];
            maxArea += currArea <= 0 ? 0 : currArea;
        }

        return maxArea;
        
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        maxArea = 0
        stack = []
        for i, h in enumerate(heights):
            start = i
            while stack and stack[-1][1] > h:
                idx, height = stack.pop()
                maxArea = max(maxArea, (i - idx) * height)
                start = idx
            stack.append((start, h))
        
        while stack:
            i, h = stack.pop()
            maxArea = max(maxArea, h * (len(heights) - i))
        return maxArea

        